### PR TITLE
Move classes to intersection object and don't emit notifications

### PR DESCRIPTION
**Pull request by TheMarex**
# Issue

Implements #4290 by including the class of the outgoing road in the intersection object.

## Tasklist
 - [x] review
 - [x] adjust for comments

### CODE REVIEW

File: features/car/classes.feature [hunk 1]
```diff
@@ -17,13 +17,13 @@ Feature: Car - Mode flag
...
         When I route I should get
-            | from | to | route       | turns                                              | classes      |
-            | a    | d  | ab,bc,cd,cd | depart,notification right,notification left,arrive | ,ferry,,     |
-            | d    | a  | cd,bc,ab,ab | depart,notification right,notification left,arrive | ,ferry,,     |
-            | c    | a  | bc,ab,ab    | depart,notification left,arrive                    | ferry,,      |
-            | d    | b  | cd,bc,bc    | depart,notification right,arrive                   | ,ferry,ferry |
-            | a    | c  | ab,bc,bc    | depart,notification right,arrive                   | ,ferry,ferry |
-            | b    | d  | bc,cd,cd    | depart,notification left,arrive                    | ferry,,      |
+            | from | to | route       | turns                                              | classes                  |
+            | a    | d  | ab,bc,cd,cd | depart,notification right,notification left,arrive | [[]],[[ferry]],[[]],[[]] |
```
Comment (MoKob) [inline features/car/classes.feature]: I like this `[]` approach. We should think about doing this on all outputs for intersections.

File: include/engine/guidance/assemble_steps.hpp [hunk 2]
```diff
@@ -119,6 +120,8 @@ inline std::vector<RouteStep> assembleSteps(const datafacade::BaseDataFacade &fa
                 const auto exits = facade.GetExitsForID(step_name_id);
                 const auto distance = leg_geometry.segment_distances[segment_index];
                 auto classes = facade.GetClasses(path_point.classes);
+                // intersections contain the classes of exiting road
+                intersection.classes = std::move(classes);
```
Comment (MoKob) [inline include/engine/guidance/assemble_steps.hpp]: just a personal preference: I'd prefer not assigning a name to `classes` if we only use it to transfer the data. i.e. `intersection.classes = face.GetClasses()`.

File: include/engine/guidance/assemble_steps.hpp [hunk 3]
```diff
@@ -196,6 +198,8 @@ inline std::vector<RouteStep> assembleSteps(const datafacade::BaseDataFacade &fa
         const auto distance = leg_geometry.segment_distances[segment_index];
         const EdgeWeight duration = segment_duration + target_duration;
         const EdgeWeight weight = segment_weight + target_weight;
+        // intersections contain the classes of exiting road
+        intersection.classes = std::move(target_classes);
```
Comment (MoKob) [inline include/engine/guidance/assemble_steps.hpp]: same here.

**Comment by MoKob** [comment]
darn. pressed that button to early: the approach here seems fine, I am unsure about some stuff for collapsing, though. I am going to collect a few links and post here.

File: features/car/classes.feature [hunk 4]
```diff
@@ -96,6 +96,6 @@ Feature: Car - Mode flag
             | cd    | primary |          |
 
         When I route I should get
-            | from | to | route    | turns                            | classes             |
-            | a    | d  | ab,bc,cd | depart,notification right,arrive | toll,motorway;toll, |
+            | from | to | route    | turns         | classes                          |
+            | a    | d  | ab,cd    | depart,arrive | [[toll],[motorway,toll],[]],[[]] |
```
Comment (oxidase) [inline features/car/classes.feature]: Just minor:  spaces after comma increase readability, 
```
[[toll],[motorway,toll],[]],[[]]
[[toll],[motorway,toll],[]], [[]]
[[toll], [motorway,toll], []], [[]]
```

**Comment by TheMarex** [comment]
@MoKob added a test for the case you mentioned, seems to be working as expected.

I don't really understand what is going on at the position that you linked. Could you please outline a test case you would want to see for testing correct behavior?

File: features/car/classes.feature (hunk 4 above)
Comment (TheMarex) [inline features/car/classes.feature]: Changed to round braces for a similar effect. 👍

**Comment by MoKob** [comment]
@TheMarex I went through the entire code. It seems all locations correctly transfer the intersections. So it is more a question of what we want in the output. If it is alright that the intersections don't line up with announcements, we should be fine.

E.g. if you collapse a turn, at some points we transfer signage (names/destinations...), since the second part is the actual turn. It looks like we keep all intersections intact. So the classes refer to the exact location the intersection is in the data, which would be fine with me. We wouldn't return a motorway class in the exact step we turn onto a motorway in some cases, or if we turn onto toll road with a sliproad, the turn itself will not feature `toll` but only the second intersection will.

In total I would consider the `keep it where it is` method preferable, since we don't mess with the data and are less likely to produce errors.

**Pull request merged**